CMake and some other smaller changes

mpilman opened the pull request:
This pull request contains many of the changes we did to our Snowflake fork. Currently these things are not properly tested as we are still in the process of building our test infrastructure.

The changes are:
- Can be built with cmake (except TLS)
- CMake can create RPM and DEB packages
- ctest will run all simulation and unit tests
- changed compiler to use c++17 (necessary for our serialization change that will come later)
- error definitions are now generated out of an xml file
- Errors will contain the location from where they were created (but these are not sent over the network). This proofed useful for debugging
- Our TPCC workload

alexmiller-apple commented:
I'm super excited to see this PR.  Skimming over the changes, there is... a lot here.  Would it be possible to do this a bit more piecewise?  From your overview I'm going to guess that everything depends on the CMake portion of this change, as without it, you don't have a way to build/test what you did.

Would it be possible to post a PR of *just* the:

> * Can be built with cmake (except TLS)
> * CMake can create RPM and DEB packages
> * ctest will run all simulation and unit tests

And then we could work on merging one PR of:

> * error definitions are now generated out of an xml file
> * Errors will contain the location from where they were created (but these are not sent over the network). This proofed useful for debugging

and the other of:

> * Our TPCC workload

?

Then we have left:

> * changed compiler to use c++17 (necessary for our serialization change that will come later)

which is going to be a bit more complicated to merge, as it's going to require updating the published docker image, and thus the supported toolchain, to something that also supports c++17.

> * 6d23441 Use abspath when dealing with the simulator file-cache

is something I had talked over with @satherton before, who seemed of the opinion that not closing files was an intentional feature, due to it then properly emulating that one can open a file, do changes, close the file, open the file again, and get a copy of the file that's different from what's persisted on disk.  I think, and Steve will probably correct me if I'm wrong.

We had implemented some flag to IAsyncFile::open() that stops a file from entering this cache that was used for backup, if that's what caused problems for you?  I don't know if that code was in 3.0 or not, and @satherton probably would?

mpilman commented:
Moving this into separate pull requests will be incredibly painful - these changes were made quite a while ago (as you will see) and merges resulted in many conflicts that I had to resolve manually (the last merge took me more than a day). As you know the reason for this was mostly legal and not technical. In general, I agree that small pull requests are much better, but in this case it would be extremely painful for us. This is why I made a pull request for all changes at once (TPCC is kind of a special case - this one can be simply removed and added later as it is a contained single file).

So if we could find any kind of middle way that would make this less painful for me, I would be very happy. I don't mind working on solutions for all problems you can find with the pull requests but if I really have to put this into several pull requests it would probably mean that the quickest way of achieving this would be to throw everything away and manually replay it (and I hope you can see why I am not very eager on doing this).

> which is going to be a bit more complicated to merge, as it's going to require updating the published docker image, and thus the supported toolchain, to something that also supports c++17.
I think this code can still be compiled with C++14. The change here affects only cmake and it makes the code working with a C++17 compiler (for example, this code replaces uses of std::auto_ptr with std::unique_ptr - something that is IMHO a good idea even if you are on C++14). I can change the cmake file to use C++14 for now, if that is what you are asking for.

> is something I had talked over with @satherton before, who seemed of the opinion that not closing files was an intentional feature, due to it then properly emulating that one can open a file, do changes, close the file, open the file again, and get a copy of the file that's different from what's persisted on disk.

I think you misunderstand this fix. The problem with the current code is that deleted files will not be closed. I did not remove the file cache (so if a file is opened several times - they will still share a file descriptor). Also files can still be opened and closed again. Without this change, the simulator will leak file descriptors. This was very problematic for us when we introduced a simulator feature (which we will port later) that runs more than one FDB cluster in the simulator - as the number of simulated machines will go up by a constant factor we quickly ran out of file descriptors when running tests.

alexmiller-apple commented:
I've cherry picked your cmake commits with minor modifications into [alexmiller-apple/foundationdb:cmake](https://github.com/alexmiller-apple/foundationdb/tree/cmake).  The diff between our CMake-related files is http://paste.ubuntu.com/p/ghPyTkwmJ6/.  (There's a few minor ordering changes that I should probably fix to help you avoid merge conflicts.)

It successfully compiles the binaries under bin/ for me (yay!), and I'm currently fighting with seeing if I can get it to work with the version of cmake that's in the docker image that we publish.  Otherwise, I'm going to need to either update the docker image, or drop the docker image as the definition of our supported build environment.

I honestly haven't actually really read through all the changes yet.  I'll probably be spending a roughly equivalent amount of time tomorrow doing an actual review.  I'll post the branch as a PR tomorrow.

> I think you misunderstand this fix.

It sounds like I likely did.  I'll keep pinging @satherton until he shows up an thinks about it for me though.

mpilman commented:
Awesome, thanks a lot. We (@atn34 to be more specific) are currently working on separating the cmake changes. So we will make a separate pull request for that. He will also set the required compiler to C++14 for now but we would like to change that in the future (`if constexpr` is so much more elegant than template specialization). If there is a lot of resistance against that change we will backboard our other changes that depend on 17.

For the Error codes, I think this is fine. I mostly ported them early because they proved to be very helpful when debugging. So as long as you are not generally opposed to the change, I am fine with waiting. Otherwise I would like to have a discussion about that.

mpilman commented:
Oh I see you already pulled out cmake - sorry, I completely missed your first message. This is cool, we are reviewing your fork right now.

mpilman commented:
Most of this either got merge by other PRs, has an open PR or will be put redone.
